syntax = "proto3";

package raft;

service RaftService {
  // Raft consensus RPCs
  rpc RequestVote(VoteRequest) returns (VoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  
  // Client-facing RPCs
  rpc AddInventory(AddInventoryRequest) returns (AddInventoryResponse);
  rpc UpdateInventory(UpdateInventoryRequest) returns (UpdateInventoryResponse);
  rpc GetInventory(GetInventoryRequest) returns (GetInventoryResponse);
  rpc GetLeaderInfo(GetLeaderRequest) returns (GetLeaderResponse);
  rpc QueryLLM(QueryLLMRequest) returns (QueryLLMResponse);
}

message VoteRequest {
  int32 term = 1;
  string candidateId = 2;
  int32 lastLogIndex = 3;
  int32 lastLogTerm = 4;
}

message VoteResponse {
  int32 term = 1;
  bool voteGranted = 2;
}

message LogEntry {
  int32 term = 1;
  int32 index = 2;
  string operation = 3;
  string product = 4;
  int32 qty_change = 5;
  int32 new_qty = 6;
  string username = 7;
  string timestamp = 8;
  string request_id = 9;
}

message AppendEntriesRequest {
  int32 term = 1;
  string leaderId = 2;
  int32 prevLogIndex = 3;
  int32 prevLogTerm = 4;
  repeated LogEntry entries = 5;
  int32 leaderCommit = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
  int32 matchIndex = 3;
}

// Client-facing messages
message AddInventoryRequest {
  string username = 1;
  string product = 2;
  int32 quantity = 3;
  string request_id = 4;
  string timestamp = 5;
}

message AddInventoryResponse {
  bool success = 1;
  string message = 2;
  string leader_hint = 3;
}

message UpdateInventoryRequest {
  string username = 1;
  string product = 2;
  int32 quantity = 3;
  string request_id = 4;
  string timestamp = 5;
}

message UpdateInventoryResponse {
  bool success = 1;
  string message = 2;
  string leader_hint = 3;
}

message GetInventoryRequest {
}

message InventoryItem {
  string product = 1;
  int32 quantity = 2;
}

message GetInventoryResponse {
  repeated InventoryItem items = 1;
}

message GetLeaderRequest {
}

message GetLeaderResponse {
  bool is_leader = 1;
  string leader_address = 2;
  int32 current_term = 3;
}

// LLM Query messages
message QueryLLMRequest {
  string query = 1;
  string username = 2;
  string request_id = 3;
}

message QueryLLMResponse {
  bool success = 1;
  string response = 2;
  string error = 3;
}